src/zarr/core/dtype/npy/common.py

```diff
@@ -0,0 +1,583 @@
+
+    Parameters
+    ----------
+    data : JSONFloat
+        The JSON float to convert.
+
+    Returns
+    -------
+    float
+        The float value.
+    """
+    match data:
+        case "NaN":
+            return float("nan")
+        case "Infinity":
+            return float("inf")
+        case "-Infinity":
+            return float("-inf")
+        case _:
+            return float(data)
+
+
+def float_from_json_v3(data: JSONFloat) -> float:
+    """
+    Convert a JSON float to a float (v3).
+
+    Parameters
+    ----------
+    data : JSONFloat
+        The JSON float to convert.
+
+    Returns
+    -------
+    float
+        The float value.
+    """
+    # todo: support the v3-specific NaN handling
+    return float_from_json_v2(data)
+
+
+def float_from_json(data: JSONFloat, *, zarr_format: ZarrFormat) -> float:
+    """
+    Convert a JSON float to a float based on zarr format.
+
+    Parameters
+    ----------
+    data : JSONFloat
+        The JSON float to convert.
+    zarr_format : ZarrFormat
+        The zarr format version.
+
+    Returns
+    -------
+    float
+        The float value.
+    """
+    if zarr_format == 2:
+        return float_from_json_v2(data)
+    else:
+        return float_from_json_v3(data)
+
+
+def bytes_from_json(data: str, *, zarr_format: ZarrFormat) -> bytes:
+    """
+    Convert a JSON string to bytes
+
+    Parameters
+    ----------
+    data : str
+        The JSON string to convert.
+    zarr_format : ZarrFormat
+        The zarr format version.
+
+    Returns
+    -------
+    bytes
+        The bytes.
+    """
+    if zarr_format == 2:
+        return base64.b64decode(data.encode("ascii"))
+    # TODO: differentiate these as needed. This is a spec question.
+    if zarr_format == 3:
+        return base64.b64decode(data.encode("ascii"))
+    raise ValueError(f"Invalid zarr format: {zarr_format}. Expected 2 or 3.")  # pragma: no cover
+
+
+def bytes_to_json(data: bytes, zarr_format: ZarrFormat) -> str:
+    """
+    Convert bytes to JSON.
+
+    Parameters
+    ----------
+    data : bytes
+        The bytes to store.
+    zarr_format : ZarrFormat
+        The zarr format version.
+
+    Returns
+    -------
+    str
+        The bytes encoded as ascii using the base64 alphabet.
+    """
+    # TODO: decide if we are going to make this implementation zarr format-specific
+    return base64.b64encode(data).decode("ascii")
+
+
+def float_to_json_v2(data: float | np.floating[Any]) -> JSONFloat:
+    """
+    Convert a float to JSON (v2).
+
+    Parameters
+    ----------
+    data : float or np.floating
+        The float value to convert.
+
+    Returns
+    -------
+    JSONFloat
+        The JSON representation of the float.
+    """
+    if np.isnan(data):
+        return "NaN"
+    elif np.isinf(data):
+        return "Infinity" if data > 0 else "-Infinity"
+    return float(data)
+
+
+def float_to_json_v3(data: float | np.floating[Any]) -> JSONFloat:
```

Comment:
I'm pretty sure that for a complete implementation of the v3 spec this needs to supply the option to the user of encoding as a string (see comment on the `from_json` method.

Reply:
It's not clear to me that we _have_ to provide users this option, but I agree that it would be nice to expose this somehow. But I think that's not going in this PR